File: DictationAssistant/SpeakControler.vb
Public Class SpeakControler
    Private ReadOnly lock As New Object '约定：RaiseEvent前必须unlock

    Public Property Speaker As ISpeaker
    Public Property WordListSource As IWordListAdapter
    Private WithEvents SpeakStateControler As ISpeakStateControler
    Private NextWordIndex As Integer = 0 '下一个播报词语索引
    Private ElapsedSeconds As Integer = 0 '已过秒数
    Private _ElapsedTimes As Integer = 0 '已报次数

    Private AutoMode As Boolean = False '（是否）自动播报

    Public Property CurrentWaitingTimeCalculator As IWaitingTimeCalculator = New FixedWaitingTime(3)
    Public Property AutoMode_TimesPerWord As Integer = 2 '每词遍数

    Private WithEvents TimerForAutoSpeaking As New Timers.Timer With {.Enabled = False, .Interval = 1000}

    '警告：任何事件都可能在一个全新的线程被触发！

    ''' <summary>
    ''' 自动播报完成
    ''' </summary>
    ''' <param name="Reason">原因，0：报完，1：手动结束</param>
    ''' <remarks></remarks>
    Event AutoSpeakingCompleted(Reason As Integer)

    ''' <summary>
    ''' 下次播报间隔被改变
    ''' </summary>
    ''' <param name="i"></param>
    ''' <param name="jiangge"></param>
    ''' <remarks></remarks>
    Event UpdateWaitingTimeInfo(i As Integer, jiangge As Integer)

    Public Event Speaking(i As Integer)
    Public Event StoppedThis(i As Integer, Completed As Boolean)

    Private Function GetWaitingTime(i As Integer) As Integer
        Return CurrentWaitingTimeCalculator.CalculateWaitingTime(WordListSource(i))
    End Function
    Private Sub TimerForAutoSpeaking_Elapsed(sender As Object, e As EventArgs) Handles TimerForAutoSpeaking.Elapsed
        Dim NeedToWait As Boolean = False
        Dim CompletedAuto As Boolean = False
        Dim WaitingToSpeakWord As Integer = 0
        Dim 剩余等待时间 As Integer = 0
        Dim NumberOfWords = WordListSource.Count
        Dim LocalNextWordIndex As Integer
        SyncLock lock
            LocalNextWordIndex = NextWordIndex
        End SyncLock
        Dim WaitingTime = GetWaitingTime(LocalNextWordIndex - 1)
        SyncLock lock
            ElapsedSeconds = ElapsedSeconds + 1 '加一秒
            If LocalNextWordIndex <= NumberOfWords Then '下一个 小于 最后项目+1 时
                剩余等待时间 = WaitingTime - ElapsedSeconds
                NeedToWait = 剩余等待时间 > 0
                If ElapsedTimes >= AutoMode_TimesPerWord Then '已报次数 >= 每词遍数 时
                    WaitingToSpeakWord = LocalNextWordIndex
                    If NeedToWait = False AndAlso LocalNextWordIndex = NumberOfWords Then
                        CompletedAuto = True
                    End If
                Else
                    WaitingToSpeakWord = LocalNextWordIndex - 1
                End If
            Else
                CompletedAuto = True
            End If
            If CompletedAuto Then
                AutoMode = False
                TimerForAutoSpeaking.Stop()
            End If
        End SyncLock
        If CompletedAuto Then
            RaiseEvent AutoSpeakingCompleted(0)
        Else
            TimerForAutoSpeaking.Start()
            If NeedToWait Then
                RaiseEvent UpdateWaitingTimeInfo(WaitingToSpeakWord, 剩余等待时间)
            Else
                Speak(WaitingToSpeakWord)
            End If
        End If
    End Sub

    Public Sub Speak(Index As Integer)
        Dim LocalSpeakStateControler As ISpeakStateControler
        Dim RaiseStoppedThis As Boolean = False
        Dim LastIndex As Integer
        Dim Word As String = WordListSource(Index)
        SyncLock lock
            TimerForAutoSpeaking.Stop()
            ElapsedSeconds = 0
            If SpeakStateControler IsNot Nothing Then
                RaiseStoppedThis = True
                LastIndex = NextWordIndex - 1
                SpeakStateControler.StopSpeak()
                SpeakStateControler = Nothing
            End If
            If NextWordIndex <> Index + 1 Then
                _ElapsedTimes = 0
                NextWordIndex = Index + 1
            End If
            LocalSpeakStateControler = Speaker.Speak(Word)
        End SyncLock
        If RaiseStoppedThis Then
            RaiseEvent StoppedThis(LastIndex, False)
        End If
        RaiseEvent Speaking(Index)
        SpeakStateControler = LocalSpeakStateControler
    End Sub
    Private Sub SpeakStateControler_PlayCompleted(sender As Object, e As EventArgs) Handles SpeakStateControler.PlayCompleted
        Dim localAutoMode = False
        Dim LastIndex As Integer
        SyncLock lock
            localAutoMode = AutoMode
            LastIndex = NextWordIndex - 1
            SpeakStateControler = Nothing
            _ElapsedTimes += 1
        End SyncLock

        RaiseEvent StoppedThis(LastIndex, True)
        If localAutoMode Then
            TimerForAutoSpeaking_Elapsed(sender, e)
        End If
    End Sub
    ''' <summary>
    ''' 取下一个词语的位置
    ''' </summary>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Public Function GetNextWordIndex() As Integer
        SyncLock lock
            Return NextWordIndex
        End SyncLock
    End Function
    ''' <summary>
    ''' 移动位置到第一个前
    ''' </summary>
    ''' <remarks></remarks>
    Public Sub ResetProgress()
        SyncLock lock
            NextWordIndex = 0
            _ElapsedTimes = 0
        End SyncLock
    End Sub
    ''' <summary>
    ''' 播报下一个并移动位置
    ''' </summary>
    ''' <remarks></remarks>
    Public Sub SpeakNext()
        Dim Index As Integer
        SyncLock lock
            Index = NextWordIndex
        End SyncLock
        Speak(Index)
    End Sub
    ''' <summary>
    ''' 播报当前词语
    ''' </summary>
    ''' <remarks></remarks>
    Public Sub SpeakAgain()
        Dim Index As Integer
        SyncLock lock
            Index = NextWordIndex - 1
        End SyncLock
        Speak(Index)
    End Sub
    ''' <summary>
    ''' 播报上一个并移动位置
    ''' </summary>
    ''' <remarks></remarks>
    Public Sub SpeakLast()
        Dim Index As Integer
        SyncLock lock
            Index = NextWordIndex - 2
        End SyncLock
        Speak(Index)
    End Sub
    Public ReadOnly Property IsAutoMode As Boolean
        Get
            SyncLock lock
                Return AutoMode
            End SyncLock
        End Get
    End Property
    Public ReadOnly Property IsSpeaking As Boolean
        Get
            SyncLock lock
                Return SpeakStateControler IsNot Nothing
            End SyncLock
        End Get
    End Property

    Public ReadOnly Property ElapsedTimes As Integer
        Get
            SyncLock lock
                Return _ElapsedTimes
            End SyncLock
        End Get
    End Property

    Public Sub StartAuto(Optional Index As Integer = 0)
        SyncLock lock
            AutoMode = True
            ElapsedSeconds = 0
        End SyncLock
        Speak(Index)
    End Sub
    Public Sub StopAuto()
        Dim RaiseStoppedThis As Boolean = False
        Dim LastIndex As Integer
        SyncLock lock
            If AutoMode = False Then Exit Sub
            AutoMode = False
            TimerForAutoSpeaking.Stop()
            If SpeakStateControler IsNot Nothing Then
                RaiseStoppedThis = True
                LastIndex = NextWordIndex - 1
                SpeakStateControler.StopSpeak()
                SpeakStateControler = Nothing
            End If
        End SyncLock
        If RaiseStoppedThis Then
            RaiseEvent StoppedThis(LastIndex, False)
        End If
        RaiseEvent AutoSpeakingCompleted(1)
    End Sub
    Public Sub PauseAuto()
        Dim RaiseStoppedThis As Boolean = False
        Dim LastIndex As Integer
        SyncLock lock
            If AutoMode = False Then Exit Sub
            TimerForAutoSpeaking.Stop()
            If SpeakStateControler IsNot Nothing Then
                RaiseStoppedThis = True
                LastIndex = NextWordIndex - 1
                SpeakStateControler.StopSpeak()
                SpeakStateControler = Nothing
            End If
        End SyncLock
        If RaiseStoppedThis Then
            RaiseEvent StoppedThis(LastIndex, False)
        End If
    End Sub
    Public Sub ResumeAuto()
        SyncLock lock
            If AutoMode = False Then Exit Sub
            TimerForAutoSpeaking.Start()
        End SyncLock
    End Sub

    Public Sub StopThis()
        Dim RaiseStoppedThis As Boolean = False
        Dim LastIndex As Integer
        SyncLock lock
            If SpeakStateControler IsNot Nothing Then
                RaiseStoppedThis = True
                LastIndex = NextWordIndex - 1
                SpeakStateControler.StopSpeak()
                SpeakStateControler = Nothing
            End If
        End SyncLock
        If RaiseStoppedThis Then
            RaiseEvent StoppedThis(LastIndex, False)
        End If
    End Sub
End Class

File: DictationAssistant/SpeakControler.cs
using DictationAssistant.Audio;
using DictationAssistant.Lyric;
using DictationAssistant.Speech;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.IO;
using System.Runtime.CompilerServices;

namespace DictationAssistant
{
    public class SpeakControler : INotifyPropertyChanged
    {
        private System.Timers.Timer TimerForAutoSpeaking;
        private readonly object lockObject = new object(); // 约定：RaiseEvent前必须unlock
        private int ElapsedSeconds = 0; // 已过秒数

        public SpeakControler()
        {
            TimerForAutoSpeaking = new System.Timers.Timer() { Enabled = false, Interval = 1000 };
            TimerForAutoSpeaking.Elapsed += TimerForAutoSpeaking_Elapsed;
        }


        private ISpeaker _speaker;
        public ISpeaker Speaker
        {
            get { return _speaker; }
            set
            {
                _speaker = value;
                PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(Speaker)));
            }
        }

        private IWordListAdapter _wordListSource;
        public IWordListAdapter WordListSource
        {
            get { return _wordListSource; }
            set
            {
                _wordListSource = value;
                PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(WordListSource)));
            }
        }

        private ISpeakStateControler _speakStateControler;
        private ISpeakStateControler SpeakStateControler
        {
            get
            {
                return _speakStateControler;
            }

            set
            {
                lock (lockObject)
                {
                    var old = _speakStateControler;
                    if (old != null)
                        _speakStateControler.PlayCompleted -= SpeakStateControler_PlayCompleted;
                    _speakStateControler = value;
                    if (value != null)
                    {
                        value.PlayCompleted += SpeakStateControler_PlayCompleted;
                    }
                    if ((old == null) ^ (value == null))
                    {
                        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(IsSpeaking)));
                    }
                }
            }
        }
        public bool IsSpeaking => SpeakStateControler != null;

        private int _nextWordIndex = 0;
        public int NextWordIndex
        {
            get { return _nextWordIndex; }
            private set
            {
                _nextWordIndex = value;
                PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(NextWordIndex)));
            }
        }

        private int _elapsedTimes = 0;
        public int ElapsedTimes
        {
            get
            {
                return _elapsedTimes;
            }
            private set
            {
                _elapsedTimes = value;
                PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(ElapsedTimes)));
            }
        }

        public bool AutoMode { get; private set; } = false;

        private IWaitingTimeCalculator _currentWaitingTimeCalculator = new FixedWaitingTime(3);
        public IWaitingTimeCalculator CurrentWaitingTimeCalculator
        {
            get { return _currentWaitingTimeCalculator; }
            set
            {
                _currentWaitingTimeCalculator = value;
                PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(CurrentWaitingTimeCalculator)));
            }
        }

        private int _autoMode_TimesPerWord = 2;

        public int AutoMode_TimesPerWord
        {
            get { return _autoMode_TimesPerWord; }
            set
            {
                _autoMode_TimesPerWord = value;
                PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(AutoMode_TimesPerWord)));
            }
        }

        private bool isPausedAutoMode = false;
        public bool IsPausedAutoMode
        {
            get { return isPausedAutoMode; }
            set
            {
                isPausedAutoMode = value;
                PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(IsPausedAutoMode)));
            }
        }

        public event PropertyChangedEventHandler PropertyChanged;



        // 警告：任何事件都可能在一个全新的线程被触发！

        public event UpdateWaitingTimeInfoEventHandler UpdateWaitingTimeInfo;

        public delegate void UpdateWaitingTimeInfoEventHandler(int i, int remind);

        private int GetWaitingTime(string word)
        {
            return CurrentWaitingTimeCalculator.CalculateWaitingTime(word);
        }
        private int GetWaitingTime(int i)
        {
            return GetWaitingTime(WordListSource[i]);
        }
        private void TimerForAutoSpeaking_Elapsed(object sender, EventArgs e)
        {
            bool needToWait = false;
            bool completedAuto = false;
            int waitToSpeakWord = 0;
            int remind = 0;
            var numberOfWords = WordListSource.Length;
            int localNextWordIndex;
            lock (lockObject)
                localNextWordIndex = NextWordIndex;
            var waitingTime = GetWaitingTime(localNextWordIndex - 1);
            lock (lockObject)
            {
                ElapsedSeconds += 1; // 加一秒
                if (localNextWordIndex <= numberOfWords)
                {
                    remind = waitingTime - ElapsedSeconds;
                    needToWait = remind > 0;
                    if (ElapsedTimes >= AutoMode_TimesPerWord)
                    {
                        waitToSpeakWord = localNextWordIndex;
                        if (needToWait == false && localNextWordIndex == numberOfWords)
                            completedAuto = true;
                    }
                    else
                        waitToSpeakWord = localNextWordIndex - 1;
                }
                else
                    completedAuto = true;
                if (completedAuto)
                {
                    AutoMode = false;
                    TimerForAutoSpeaking.Stop();
                }
            }
            if (completedAuto)
            {
                PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(AutoMode)));
            }
            else
            {
                TimerForAutoSpeaking.Start();
                if (needToWait)
                    UpdateWaitingTimeInfo?.Invoke(waitToSpeakWord, remind);
                else
                    Speak(waitToSpeakWord);
            }
        }

        public void Speak(int Index)
        {
            ISpeakStateControler LocalSpeakStateControler;
            string word = WordListSource[Index];
            lock (lockObject)
            {
                TimerForAutoSpeaking.Stop();
                ElapsedSeconds = 0;
                if (SpeakStateControler != null)
                {
                    SpeakStateControler.StopSpeak();
                    SpeakStateControler = null;
                }
                if (NextWordIndex != Index + 1)
                {
                    ElapsedTimes = 0;
                    NextWordIndex = Index + 1;
                }
                if (IsPausedAutoMode)
                    IsPausedAutoMode = false;
                LocalSpeakStateControler = Speaker.Speak(word);
            }
            SpeakStateControler = LocalSpeakStateControler;
        }
        private void SpeakStateControler_PlayCompleted(object sender, EventArgs e)
        {
            var localAutoMode = false;
            lock (lockObject)
            {
                localAutoMode = AutoMode;
                SpeakStateControler = null;
                ElapsedTimes += 1;
            }
            if (localAutoMode)
                TimerForAutoSpeaking_Elapsed(sender, e);
        }

        /// <summary>
        /// 移动位置到第一个前
        /// </summary>
        /// <remarks></remarks>
        public void ResetProgress()
        {
            lock (lockObject)
            {
                NextWordIndex = 0;
                ElapsedTimes = 0;
            }
        }
        /// <summary>
        /// 播报下一个并移动位置
        /// </summary>
        /// <remarks></remarks>
        public void SpeakNext()
        {
            int index;
            lock (lockObject)
                index = NextWordIndex;
            Speak(index);
        }
        /// <summary>
        /// 播报当前词语
        /// </summary>
        /// <remarks></remarks>
        public void SpeakAgain()
        {
            int index;
            lock (lockObject)
                index = NextWordIndex - 1;
            Speak(index);
        }
        /// <summary>
        /// 播报上一个并移动位置
        /// </summary>
        /// <remarks></remarks>
        public void SpeakPrevious()
        {
            int index;
            lock (lockObject)
                index = NextWordIndex - 2;
            Speak(index);
        }

        public void StartAuto(int index = 0)
        {
            lock (lockObject)
            {
                if (IsPausedAutoMode)
                    IsPausedAutoMode = false;
                AutoMode = true;
                ElapsedSeconds = 0;
            }
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(AutoMode)));
            Speak(index);
        }
        public void StopAuto()
        {
            lock (lockObject)
            {
                if (AutoMode == false)
                    return;
                if (IsPausedAutoMode)
                    IsPausedAutoMode = false;
                AutoMode = false;
                TimerForAutoSpeaking.Stop();
                if (SpeakStateControler != null)
                {
                    SpeakStateControler.StopSpeak();
                    SpeakStateControler = null;
                }
            }
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(AutoMode)));
        }

        public void PauseAuto()
        {
            lock (lockObject)
            {
                if (AutoMode == false)
                    return;
                TimerForAutoSpeaking.Stop();
                if (SpeakStateControler != null)
                {
                    SpeakStateControler.StopSpeak();
                    SpeakStateControler = null;
                }
                IsPausedAutoMode = true;
            }
        }

        public void ResumeAuto()
        {
            lock (lockObject)
            {
                if (AutoMode == false)
                    return;
                TimerForAutoSpeaking.Start();
                IsPausedAutoMode = false;
            }
        }

        public void StopThis()
        {
            lock (lockObject)
            {
                if (SpeakStateControler != null)
                {
                    SpeakStateControler.StopSpeak();
                    SpeakStateControler = null;
                }
            }
        }

        public bool SaveAudio(
            IVoice voice,
            SpeakParam speakParam,
            PcmWriter pcmWriter,
            ILyricWriter lrcWriter,
            Action<double> reportProgress = null,
            Func<bool> isCancellationPending = null)
        {
            bool complete = true;

            reportProgress?.Invoke(0);
            long byteOffest = 0;
            int length = WordListSource.Length;
            string[] wordList = new string[length];
            for (int i = 0; i < length; i++)
            {
                wordList[i] = WordListSource[i];
            }

            PcmStreamWithInfo speech;
            for (int iWord = 0; iWord < length; iWord++)
            {
                if (isCancellationPending != null && isCancellationPending())
                {
                    complete = false;
                    break;
                }
                reportProgress?.Invoke((double)iWord / length);
                string word = wordList[iWord];

                if (AutoMode_TimesPerWord >= 1)
                {
                    speech = voice.Speak(word, speakParam);
                    lrcWriter?.Append(word, pcmWriter.BytesToMilliseconds(byteOffest));
                    byteOffest += pcmWriter.Write(speech);
                    byteOffest += pcmWriter.WriteDelay(GetWaitingTime(word) * 1000);
                    for (var i = 2; i <= AutoMode_TimesPerWord; i++)
                    {
                        if (speech.PcmStream.CanSeek)
                        {
                            speech.PcmStream.Seek(0, SeekOrigin.Begin);
                        }
                        else
                        {
                            speech.Dispose();
                            speech = voice.Speak(word, speakParam);
                        }
                        lrcWriter?.Append(word, pcmWriter.BytesToMilliseconds(byteOffest));
                        byteOffest += pcmWriter.Write(speech);
                        byteOffest += pcmWriter.WriteDelay(GetWaitingTime(word) * 1000);
                    }
                    speech.Dispose();
                }

                lrcWriter?.Append(" ", pcmWriter.BytesToMilliseconds(byteOffest) - 10);
            }
            lrcWriter?.Append("本文件由 自动默写 程序自动生成", pcmWriter.BytesToMilliseconds(byteOffest));
            reportProgress?.Invoke(1);

            return complete;
        }
    }
}
